Mage.Sets/src/mage/cards/a/AshcoatOfTheShadowSwarm.java

```diff
@@ -0,0 +1,101 @@
+package mage.cards.a;
+
+import mage.MageInt;
+import mage.abilities.Ability;
+import mage.abilities.common.AttacksOrBlocksTriggeredAbility;
+import mage.abilities.common.BeginningOfEndStepTriggeredAbility;
+import mage.abilities.costs.Cost;
+import mage.abilities.costs.CostImpl;
+import mage.abilities.dynamicvalue.DynamicValue;
+import mage.abilities.dynamicvalue.common.PermanentsOnBattlefieldCount;
+import mage.abilities.effects.common.DoIfCostPaid;
+import mage.abilities.effects.common.ReturnFromGraveyardToHandTargetEffect;
+import mage.abilities.effects.common.continuous.BoostControlledEffect;
+import mage.cards.CardImpl;
+import mage.cards.CardSetInfo;
+import mage.constants.*;
+import mage.filter.common.FilterCreatureCard;
+import mage.filter.common.FilterCreaturePermanent;
+import mage.game.Game;
+import mage.players.Player;
+import mage.target.common.TargetCardInYourGraveyard;
+
+import java.util.UUID;
+
+/**
+ * @author PurpleCrowbar
+ */
+public final class AshcoatOfTheShadowSwarm extends CardImpl {
+
+    private static final FilterCreaturePermanent filter = new FilterCreaturePermanent(SubType.RAT, "Rats");
+    private static final FilterCreatureCard filter2 = new FilterCreatureCard("Rat creature cards");
+
+    static {
+        filter2.add(SubType.ZOMBIE.getPredicate());
+    }
+
+    public AshcoatOfTheShadowSwarm(UUID ownerId, CardSetInfo setInfo) {
+        super(ownerId, setInfo, new CardType[]{CardType.CREATURE}, "{3}{B}");
+        addSuperType(SuperType.LEGENDARY);
+        this.subtype.add(SubType.RAT, SubType.WARLOCK);
+        this.power = new MageInt(3);
+        this.toughness = new MageInt(4);
+
+        // Whenever Ashcoat of the Shadow Swarm attacks or blocks, other Rats you control
+        // get +X/+X until end of turn, where X is the number of Rats you control.
+        DynamicValue xValue = new PermanentsOnBattlefieldCount(filter);
+        this.addAbility(new AttacksOrBlocksTriggeredAbility(new BoostControlledEffect(
+                xValue, xValue, Duration.EndOfTurn, filter, true
+        ).setText("other Rats you control get +X/+X until end of turn, where X is the number of Rats you control"), false));
+
+        // At the beginning of your end step, you may mill four cards. If you do,
+        // return up to two Rat creature cards from your graveyard to your hand.
+        Ability ability = new BeginningOfEndStepTriggeredAbility(new DoIfCostPaid(
+                new ReturnFromGraveyardToHandTargetEffect(),
+                new AshcoatCost()
+        ), TargetController.YOU, false);
+        ability.addTarget(new TargetCardInYourGraveyard(0, 2, filter2));
```

Comment:
I changed the code to fix this and it is now improved in that it lets you choose the rats in the graveyard *after* milling, but I'm now running into the issue of rat cards milled by the first part of the ability not being shown as available options in the second part of the ability. See below, where red lined cards were in the graveyard before milling, and pink lined cards were put into the graveyard as a result of milling (note the unselectable rat card):
![image](https://user-images.githubusercontent.com/26198472/203678439-5d358a02-d7bb-4311-a1a1-70d1c1c9a884.png)
I guess it's because the effects are taking place at the same time so the second effect doesn't count cards milled by the first effect, but not sure. Should I just get rid of the `MillCardsControllerEffect` and include the self-milling in the custom effect?

Reply:
Nevermind, I just made a typo in the filter. This should be good to go now